DOC: Enhance description/usage for np.linalg.eig*h

stuartarchibald opened the pull request:
Proposed documentation enhancement to the `np.linalg.eig*h`
routines.

The current documentation does not reflect the nature of the
underlying LAPACK routines and the way in which they interpet
the input. This documentation and examples are with view of
adding clarity to what is actually performed.

stuartarchibald committed:
DOC: Enhance description/usage for np.linalg.eig*h

Proposed documentation enhancement to the `np.linalg.eig*h`
routines.

The current documentation does not reflect the nature of the
underlying LAPACK routines and the way in which they interpet
the input. This documentation and examples are with view of
adding clarity to what is actually performed.

wrwrwr commented:
These functions seem to give hardly explainable results if the matrix is non-symmetric, for instance:

```python
eigvalsh([[0, 1], [2, 3]])
# array([-1.,  4.])
```

Hence, wouldn't it be better to either:
* add a check and raise an exception for non-symmetric / non-Hermitian matrices;
* add an explicit note that in the above cases the result is undefined.

charris commented:
@wrwrwr I think the result is well defined, just possibly unexpected. The use of the upper/lower triangular part avoids the numerical difficulty of determining true symmetry and is also compatible with packed storage of Hermitean arrays.

charris commented on numpy/linalg/linalg.py:
```diff
@@ -930,7 +930,11 @@ def eigvalsh(a, UPLO='L'):
     UPLO : {'L', 'U'}, optional
         Same as `lower`, with 'L' for lower and 'U' for upper triangular.
-        Deprecated.
+        Deprecated. 
```
Hmm, this doesn't look deprecated in the code. @pv Is this a stray bit left over from the merge of the gufuncs?

pv replied:
I have no recollection of this, but 74e14775a144794 mentions `lower` so maybe the idea was to harmonize this with scipy's kwargs and replace the U/L with lower=True/False kwarg.

pv replied:
cf signature https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.linalg.eigvalsh.html#scipy.linalg.eigvalsh

charris replied:
@pv, OK. I'll remove it until such time, if ever, it gets harmonized with scipy ;) Speaking of which, we might should do the same for the `qr` decomposition at some point.

charris committed:
DOC: Remove "deprecated" from eigvalsh UPLO.

The UPLO parameter of eigvalsh was marked as deprecated, apparently in anticipation
of convergence with the scipy version which uses a `lower` boolean parameter. There is
no deprecation in the code, nor a lower parameter, so remove the deprecation.

[ci skip]

charris commented:
Thanks @stuartarchibald .
